File: homework/tai_nguyen/lesson7/337_house_robber_iii.ts
/**
 * @link https://leetcode.com/problems/house-robber-iii/
 */

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function isLeaf(node: TreeNode) {
  return node.left === null && node.right === null;
}

function rob(root: TreeNode | null): number {
  const cache = new Map<TreeNode, number>();

  function robWithCache(node: TreeNode | null | undefined): number {
    if (node === null || node === undefined) return 0;

    if (cache.has(node)) {
      return cache.get(node)!;
    }

    if (isLeaf(node)) {
      cache.set(node, node.val);
      return node.val;
    }

    const amount = Math.max(
      /* choose root, skip children and leverage the max to the grandchildren */
      node.val +
        robWithCache(node.left?.left) +
        robWithCache(node.left?.right) +
        robWithCache(node.right?.left) +
        robWithCache(node.right?.right),

      /* do not choose root, so we leverage to it's children  */
      robWithCache(node.left) + robWithCache(node.right)
    );

    cache.set(node, amount);
    return amount;
  }

  return robWithCache(root);
}

File: homework/thinhbka/lesson-5-graph/337. House Robber III.py
# https://leetcode.com/problems/house-robber-iii/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def dfs(self,root):
        if not root:return 0
        if root in self.hash_map:
            return self.hash_map[root]
        if root.left is None and root.right is None:
            self.hash_map[root] = root.val
            return root.val
        ans = root.val
        if root.left:
            node = root.left
            ans += self.dfs(node.left) + self.dfs(node.right)
        if root.right:
            node = root.right
            ans+=self.dfs(node.left) + self.dfs(node.right)
        ans1 = self.dfs(root.left) + self.dfs(root.right)
        self.hash_map[root] = max(ans,ans1)
        return self.hash_map[root]
        
    def rob(self, root: Optional[TreeNode]) -> int:
        self.hash_map = {}
        return self.dfs(root)